Дз по AOP

Greyzl opened the pull request.

Commit:
Дз по AOP

stvort commented on hw05-proxy/src/main/java/ru/otus/aop/logging/LoggingHandler.java:
```diff
@@ -0,0 +1,51 @@
+    }
+    @Override
+    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+        Class<?> calculatorServiceClass = calculatorService.getClass();
+        Method[] calculatorServiceMethods = calculatorServiceClass.getDeclaredMethods();
+        for (Method calculatorServiceMethod : calculatorServiceMethods){
+            if (isSameMethods(method, calculatorServiceMethod) &&
+                    calculatorServiceMethod.getDeclaredAnnotation(Log.class) != null){
+                StringBuilder builder = new StringBuilder();
+                for (Object object: args){
+                    builder.append(object).append(" ");
+                }
+                System.out.println("executed method: " + calculatorServiceMethod.getName()
+                        + ", params:" + builder);
+            }
+        }
```
Для каждого конкретного класса, аннотированные методы будут всегда одни и те же, нет нужды разбирать класс рефлексией при всех вызовах `invoke`. Это можно сделать один раз

stvort commented on hw05-proxy/src/main/java/ru/otus/aop/logging/LoggingHandler.java:
```diff
@@ -0,0 +1,51 @@
+        Class<?>[] objectTestMethodParamTypes = proxiedObjectMethod.getParameterTypes();
+        for (int paramIndex = 0; paramIndex < invokedMethod.getParameterCount(); paramIndex++){
+            if (!objectTestMethodParamTypes[paramIndex].getName().equals(
+                    proxyMethodParamTypes[paramIndex].getName())){
```
Можно сравнивать сами классы типов

Commit:
1. Добавил сохранение методов с аннотациями при создании хэндлера чтобы не рефлексировать
2. Заменил проверку имен типов на проверку самих типов в сравнении методов
3. Добавил форматтер для логирования

stvort commented on hw05-proxy/src/main/java/ru/otus/aop/logging/LoggingHandler.java:
```diff
@@ -40,8 +59,7 @@ private Boolean isSameMethods(Method invokedMethod, Method proxiedObjectMethod){
+            if (objectTestMethodParamTypes[paramIndex] != proxyMethodParamTypes[paramIndex]){
                 sameSignature = false;
                 break;
             }
```
`Arrays.equals`

stvort commented on hw05-proxy/src/main/java/ru/otus/aop/logging/LoggingHandler.java:
```diff
@@ -1,36 +1,55 @@
         return method.invoke(calculatorService,args);
     }
 
+    private Set<Method> getLogAnnotatedMethods(Object object){
+        Class<?> objectClass = object.getClass();
+        Method[] objectClassDeclaredMethods = objectClass.getDeclaredMethods();
+        Set<Method> annotatedMethods = new HashSet<>();
+        for (Method objectClassMethod : objectClassDeclaredMethods){
+            if (objectClassMethod.getDeclaredAnnotation(Log.class) != null){
+                annotatedMethods.add(objectClassMethod);
+            }
+        }
+        return annotatedMethods;
+    }
```
Через стримы короче получится

stvort commented on hw05-proxy/src/main/java/ru/otus/aop/logging/LoggingHandler.java:
```diff
@@ -1,36 +1,55 @@
+        Method[] objectClassDeclaredMethods = objectClass.getDeclaredMethods();
+        Set<Method> annotatedMethods = new HashSet<>();
+        for (Method objectClassMethod : objectClassDeclaredMethods){
+            if (objectClassMethod.getDeclaredAnnotation(Log.class) != null){
```
`isAnnotationPresent`

stvort commented on hw05-proxy/src/main/java/ru/otus/aop/logging/LoggingHandler.java:
```diff
@@ -1,36 +1,55 @@
 
+    private final LogMethodFormatter logMethodFormatter = new LogMethodFormatter();
+
+    private final Set<Method> loggingMethods;
```
Если сделать структуру с методом и его аргументами, то можно будет применять `contains`

Commit:
1. Поправил замечания в LoggingHandler.java
